README.md
# RefParse [![Current Version](https://img.shields.io/badge/version-0.1.1-green.svg)](https://github.com/peterhs73/RefParse) [![](https://img.shields.io/badge/python-3.6+-blue.svg)](https://www.python.org/downloads/) [![License](https://img.shields.io/badge/License-BSD%202--Clause-orange.svg)](https://opensource.org/licenses/BSD-2-Clause)

The RefParse is a simple citation generation tool for doi and arXiv articles. It provides both a graphical and a command-line interface.

The API is a combination of the crossref REST API and arXiv API. The format template is in Cheetah3 format. See [RefParse wiki](https://github.com/peterhs73/RefParse/wiki) for more detailed description on API, user interface, and custom format templates.

![RefParse GUI](https://media.giphy.com/media/KVRz3bp64jxjN8HJis/giphy.gif)

## Quickstart

Installation from the newest release v0.1.1

	python -m pip install git+https://github.com/peterhs73/RefParse.git@v0.1.1#egg=RefParse


### Develop

Clone the package

    git clone https://github.com/peterhs73/RefParse.git

under the RefParse directory, install the package in editing mode

    pip install . -e

run tests

	tox

## Features
- Automatically determine if link is doi or arXiv ID
- Allow user customization for formatting
- Comes with custom filters for data handling

See wiki for more details

## Reference

- [RefParse wiki](https://github.com/peterhs73/RefParse/wiki)
- [Change Log](https://github.com/peterhs73/RefParse/blob/master/CHANGELOG.md)
- [Releases](https://github.com/peterhs73/RefParse/releases)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [0.1.1] - 2021-02-09
### Added
- Add BSD license
- Add MANIFEST.in for packaging
- Reformatted README.md

### Changed
- Change gui log section size to half of its sizehint
- Change gui gif link

### Fixed
- Fix entry points value that causes namespace issue

## [0.1.0] - 2021-02-01

Initiate the project.

### Added

- Add functionality in command line
- Add template for default bib, markdown, rst and text format
- Add package level logging
- Add GUI interface
- Add parser for crossref (doi) and arXiv
- Add unittest
- Add user configuration


tests/test_api.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


from refparse.api import RefAPI
from unittest.mock import patch, Mock
import logging
import os
import yaml

curpath = os.path.dirname(os.path.realpath(__file__))
with open(os.path.join(curpath, "..", "refparse", "config.yaml"), "r") as f:
    CONFIG = yaml.load(f, Loader=yaml.SafeLoader)


@patch("refparse.parser.requests.get")
def test_incorrect_reference(mock_get, caplog):
    """Test the RefAPI class when the reference is incorrect"""
    mock_get.return_value.ok = False

    RefAPI("random/url", {})
    assert caplog.record_tuples == [
        ("API", logging.ERROR, "random/url is not a valid doi or arXiv ID")
    ]


@patch("refparse.api.RefAPI.__init__", return_value=None)
def test_reference_matching(mock_refapi_init, caplog):
    """Test the reference_matching method for RefAPI"""

    assert RefAPI().match_reference("arXiv:hep-th/9901001v3") == (
        "hep-th/9901001v3",
        "arXiv",
    )
    assert RefAPI().match_reference("arXiv:1807.01219") == (
        "1807.01219",
        "arXiv",
    )
    assert RefAPI().match_reference(
        "http://dx.doi.org/10.1021/acs.jpcc.8b11783"
    ) == ("10.1021/acs.jpcc.8b11783", "crossref")

    assert RefAPI().match_reference("test/url") == ("test/url", "")
    assert caplog.record_tuples == [
        ("API", 40, "test/url is not a valid doi or arXiv ID")
    ]


BIBTEX = """@Article{Shi2010jul
  author = {Shi, Guanming and Chavas, Jean-paul and Stiegert, Kyle},
  title = {An Analysis of the Pricing of Traits in the U.S. Corn Seed Market},
  journal = {American Journal of Agricultural Economics},
  year = {2010},
  month = {jul},
  volume = {92},
  issue = {5},
  doi = {10.1093/ajae/aaq063},
  url = {http://doi.org/10.1093/ajae/aaq063},
  pages = {1324--1338}
  abstract = {}
}
"""

MD = (
    "[^Shi2010jul]: **Shi2010jul** Shi, Guanming., Chavas, Jean-paul., "
    'Stiegert, Kyle. "An Analysis of the Pricing of Traits in the U.S. '
    'Corn Seed Market". *American Journal of Agricultural Economics*, '
    "**2010**, *92*, 1324--1338 "
    "[10.1093/ajae/aaq063](http://doi.org/10.1093/ajae/aaq063).\n"
)

RST = (
    "..[#Shi2010jul] [**Shi2010jul**] Shi, Guanming., Chavas, Jean-paul., "
    'Stiegert, Kyle. "An Analysis of the Pricing of Traits in the U.S. '
    'Corn Seed Market". *American Journal of Agricultural Economics*, '
    "**2010**, *92*, 1324--1338 "
    "[`10.1093/ajae/aaq063 <http://doi.org/10.1093/ajae/aaq063>`__]\n"
)


TEXT = (
    "Shi2010jul Shi, Guanming., Chavas, Jean-paul., Stiegert, Kyle. "
    '"An Analysis of the Pricing of Traits in the U.S. Corn Seed Market". '
    "American Journal of Agricultural Economics, 2010, 92, 1324--1338 "
    "doi:10.1093/ajae/aaq063\n"
)


@patch("refparse.api.RefAPI.__init__", return_value=None)
def test_render_failure(mock_refapi_init, caplog):
    """Test the render method for RefAPI if the status is false"""
    api_obj = RefAPI()
    api_obj.status = False
    assert api_obj.render("") is None


@patch("refparse.api.RefAPI.__init__", return_value=None)
def test_render_default(mock_refapi_init, caplog):
    """Test the render method for RefAPI if the status is false"""
    api_obj = RefAPI()
    api_obj.status = True
    api_obj.format_template = CONFIG
    api_dict = {
        "doi": "10.1093/ajae/aaq063",
        "reference": "10.1093/ajae/aaq063",
        "ref_type": "doi",
        "url": "http://doi.org/10.1093/ajae/aaq063",
        "has_publication": True,
        "journal_full_title": "American Journal of Agricultural Economics",
        "journal_abbrev_title": "American Journal of Agricultural Economics",
        "author": [
            ["Shi", "Guanming"],
            ["Chavas", "Jean-paul"],
            ["Stiegert", "Kyle"],
        ],
        "title": (
            "An Analysis of the Pricing of "
            "Traits in the U.S. Corn Seed Market"
        ),
        "title_latex": (
            "An Analysis of the Pricing of "
            "Traits in the U.S. Corn Seed Market"
        ),
        "title_html": (
            "An Analysis of the Pricing of "
            "Traits in the U.S. Corn Seed Market"
        ),
        "abstract": "",
        "online_year": "2010",
        "online_month": "07",
        "online_day": "19",
        "has_print": True,
        "print_year": "2010",
        "print_month": "07",
        "print_day": "19",
        "pages": ["1324", "1338"],
        "volume": "92",
        "issue": "5",
    }
    api_obj.output = {}
    api_obj.parser = Mock(parsed=api_dict)

    assert api_obj.render("bibtex") == BIBTEX
    assert api_obj.render("md") == MD
    assert api_obj.render("rst") == RST
    assert api_obj.render("text") == TEXT


tests/test_utils.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from refparse import utils
from unittest.mock import Mock
from bs4 import BeautifulSoup


def test_filter_month_abbr():
    """Test Filter.month_abb"""

    assert utils.Filters.month_abbr("") == ""
    assert utils.Filters.month_abbr("01") == "Jan"
    assert utils.Filters.month_abbr(1) == "Jan"


def test_filter_month_name():
    """Test Filter.month_abb"""

    assert utils.Filters.month_name("") == ""
    assert utils.Filters.month_name("01") == "January"
    assert utils.Filters.month_name(1) == "January"


def test_get_attr():
    """Test get_attr for nested attributes"""
    nested = Mock(attr1="test", spec="attr1")
    test_obj = Mock(attr2=nested, spec="attr2")

    assert utils.get_attr(test_obj, "attr2/attr1") == "test"
    assert utils.get_attr(test_obj, "attr1/attr3").string == ""


def test_get_string():
    """Test get_attr for nested attributes"""
    html = "<body><test>   Hello world \n </test></body>"
    soup = BeautifulSoup(html, 'xml')
    assert utils.get_string(soup, "body/test") == "Hello world"


def test_empty():
    """Test the Empty class"""
    empty = utils.Empty()
    assert empty.string == ""
    assert empty.get_text(strip=True) == ""


def test_html_convert():
    """Test the html_covert function"""
    html_string = """<body><b>hello</b>
    world<sub>2</sub>
    <random>!</random>
    </body>
    """
    soup = BeautifulSoup(html_string, "xml").body
    plain, latex, html = utils.html_convert(soup)
    assert plain == "hello world2 !"
    assert latex == "\\textbf{hello} world\\textsubscript{2} !"
    assert html == "<b>hello</b> world<sub>2</sub> <random>!</random>"


tests/test_parser.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from unittest.mock import patch
from refparse.parser import CrossRefParser, arXivParser, ParserBase
import os
import logging

curpath = os.path.dirname(os.path.realpath(__file__))
with open(os.path.join(curpath, "arXiv_test_example.xml"), "r") as f:
    ARXIV_XML = f.read()
with open(os.path.join(curpath, "crossref_test_example.xml"), "r") as f:
    DOI_XML = f.read()


class TestParser(ParserBase):
    __test__ = False
    """Empty parser to test on"""

    REFNAME = "TEST ID"
    QUERY_URL = "http://{}"
    REF_URL = "http://{}"
    HEADER = {}

    def parse_api(self):
        return {}


@patch("refparse.parser.requests.get")
def test_parser_false_status(mock_get, caplog):
    """Test response values and status

    A mock parser is crated, which does not parse the result
    """

    mock_get.return_value.text = ""
    mock_get.return_value.ok = False
    mock_get.return_value.status_code = 400
    TestParser("reference")

    mock_get.return_value.status_code = 504
    TestParser("reference")

    assert caplog.record_tuples == [
        ("TestParser", logging.ERROR, "Incorrect TEST ID"),
        ("TestParser", logging.ERROR, "Gateway timeout, please try again"),
    ]


# Test arXiv parser
@patch("refparse.parser.requests.get")
def test_arXiv_parser(mock_get, caplog):
    """Test arXiv parser with arXiv:hep-th/9901001v3"""
    mock_get.return_value.ok = True
    mock_get.return_value.text = ARXIV_XML

    parser = arXivParser("hep-th/9901001v3")

    api_dict = {
        "arXiv ID": "hep-th/9901001v3",
        "reference": "hep-th/9901001v3",
        "ref_type": "arXiv_ID",
        "url": "http://arxiv.org/hep-th/9901001v3",
        "has_publication": False,
        "has_print": False,
        "abstract": "We explicitly give the correspondence between spectra of "
        "heterotic string theory compactified on $T^2$ and string "
        "junctions in type IIB theory compactified on $S^2$.",
        "title": "String Junctions and Their Duals in Heterotic String Theory",
        "title_latex": "String Junctions and Their "
        "Duals in Heterotic String Theory",
        "online_year": "1999",
        "online_month": "5",
        "online_day": "10",
        "author": [["Imamura", "Yosuke"]],
    }

    assert parser.parsed == api_dict
    assert caplog.record_tuples == [
        (
            "arXivParser",
            logging.WARNING,
            "article has doi: http://dx.doi.org/10.1143/PTP.101.1155",
        )
    ]


@patch("refparse.parser.requests.get")
def test_crossref_parser(mock_get, caplog):
    """Test parsed CrossrefParser with doi: 10.1021/acs.jpcc.8b11783"""
    mock_get.return_value.ok = True
    mock_get.return_value.text = DOI_XML

    parser = CrossRefParser("10.1021/acs.jpcc.8b11783")

    api_dict = {
        "doi": "10.1021/acs.jpcc.8b11783",
        "reference": "10.1021/acs.jpcc.8b11783",
        "ref_type": "doi",
        "url": "http://doi.org/10.1021/acs.jpcc.8b11783",
        "has_publication": True,
        "journal_full_title": "The Journal of Physical Chemistry C",
        "journal_abbrev_title": "J. Phys. Chem. C",
        "author": [
            ["Tirmzi", "Ali Moeed"],
            ["Christians", "Jeffrey A."],
            ["Dwyer", "Ryan P."],
            ["Moore", "David T."],
            ["Marohn", "John A."],
        ],
        "title": (
            "Substrate-Dependent Photoconductivity Dynamics "
            "in a High-Efficiency Hybrid Perovskite Alloy"
        ),
        "title_latex": (
            "Substrate-Dependent Photoconductivity Dynamics "
            "in a High-Efficiency Hybrid Perovskite Alloy"
        ),
        "title_html": (
            "Substrate-Dependent Photoconductivity Dynamics "
            "in a High-Efficiency Hybrid Perovskite Alloy"
        ),
        "abstract": "",
        "online_year": "2019",
        "online_month": "01",
        "online_day": "17",
        "has_print": True,
        "print_year": "2019",
        "print_month": "01",
        "print_day": "17",
        "pages": ["3402", "3415"],
        "volume": "123",
        "issue": "6",
    }

    assert parser.parsed == api_dict


refparse/api.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""Main API class"""


from refparse.parser import CrossRefParser, arXivParser
from Cheetah.Template import Template
from refparse.utils import Filters
import logging
import re

api_logger = logging.getLogger("API")
api_method = {"crossref": CrossRefParser, "arXiv": arXivParser}


class RefAPI:
    """Abstract base class for user interaction

    For attribute that is None, empty string will be returned
    """

    def __init__(self, reference, format_template):
        """Initiate the object with different apis

        :param url string: the url of the target
        """

        cleaned_ref, api_type = self.match_reference(reference)
        if api_type:
            self.format_template = format_template
            self.parser = api_method[api_type](cleaned_ref)
            self.status = self.parser.ok
            self.output = {}
        else:
            self.status = False

    def match_reference(self, reference):
        """Match reference into either doi or arXiv ID

        The pattern for doi can be found on the API page
        arXiv ID has types, pre-2007 and post-2007
        Here the patterns are slightly modified to do a full search
        """
        doi_pattern = re.compile(r"10.\d{4,9}/[-._;()/:a-zA-Z0-9]+")
        arxiv_pattern1 = re.compile(r"\d{4}.\d{4,5}(v\d)?")
        arxiv_pattern2 = re.compile(r"[-a-z]+(.[A-Z]{2})?/\d{7}(v\d)?")

        if doi_pattern.search(reference):
            return doi_pattern.search(reference).group(0), "crossref"
        elif arxiv_pattern1.search(reference):
            return arxiv_pattern1.search(reference).group(0), "arXiv"
        elif arxiv_pattern2.search(reference):
            return arxiv_pattern2.search(reference).group(0), "arXiv"
        else:
            api_logger.error(f"{reference} is not a valid doi or arXiv ID")
            return reference, ""

    def render(self, ref_format):
        """Render the desired format

        If the ref_format is already rendered, it stores in the dictionary
        else it will render the format
        :param ref_format string: reference format
        """
        if not self.status:
            return
        elif ref_format not in self.output:

            template = self.format_template[ref_format]
            result = Template(
                template, searchList=[{"FN": Filters}, self.parser.parsed],
            )

            self.output[ref_format] = str(result)
        return self.output[ref_format]


refparse/gui.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""Main GUI for RefParse"""


import sys

from PySide2.QtWidgets import (
    QApplication,
    QLabel,
    QWidget,
    QPushButton,
    QRadioButton,
    QLineEdit,
    QTextEdit,
    QGridLayout,
    QHBoxLayout,
    QVBoxLayout,
    QSizePolicy,
    QShortcut,
    QButtonGroup,
)
from PySide2.QtGui import QKeySequence, QFont
from PySide2.QtCore import Slot, Signal, QThread, Qt

from refparse.api import RefAPI
import logging
from collections import defaultdict

root_logger = logging.getLogger()
gui_logger = logging.getLogger("GUI")


class ParserGUI(QWidget):
    def __init__(self, format_config):
        """The main GUI of the ref parser

        The class take output formats as an input
        """
        super().__init__()

        self.resize(800, 600)
        self.setWindowTitle("RefParse")
        self.init_layout(format_config)
        self.format_config = format_config
        self.api_object = None
        self.__threads = []

    def init_layout(self, format_config):
        """Initiate layouts

        For simplicity, the widget uses gridlayout with 2 columns
        """
        grid = QGridLayout()

        # - reference line
        self.ref_line = QLineEdit(self)
        grid.addWidget(QLabel("doi/arXiv:"), 0, 0)
        grid.addWidget(self.ref_line, 0, 1)

        # - search button
        self.search_btn = QPushButton("Search")
        self.search_btn.setShortcut("Return")
        self.search_btn.clicked.connect(self.access_reference)
        search = QHBoxLayout()
        search.addStretch(1)
        search.addWidget(self.search_btn)
        grid.addLayout(search, 1, 1)

        # - Output
        grid.addWidget(QLabel("Output:"), 2, 0)

        # - output format, set the first button to be true
        format_opt = QVBoxLayout()
        self.format_btns = QButtonGroup(self)
        self.format_btns.buttonClicked.connect(self.change_format)
        for format_type in format_config.keys():
            btn = QRadioButton(format_type, self)
            format_opt.addWidget(btn)
            self.format_btns.addButton(btn)
        # set first button checked
        self.format_btns.button(-2).setChecked(True)
        self.output_box = QTextEdit(self)

        grid.addLayout(format_opt, 3, 0, Qt.AlignTop)
        grid.addWidget(self.output_box, 3, 1)

        # - copy button
        self.copy_btn = QPushButton("Copy")
        self.copy_btn.clicked.connect(self.copy)
        copy = QHBoxLayout()
        copy.addStretch(1)
        copy.addWidget(self.copy_btn)
        grid.addLayout(copy, 4, 1)

        # - log box (use custom logging handler that stream log to text box)
        self.log_box = QTextEdit(self)
        self.log_box.setReadOnly(True)
        self.log_box.setStyleSheet("background-color: transparent;")
        # set the log box to half of the size
        self.log_box.setMaximumHeight(self.log_box.sizeHint().height() / 2)

        # custom signal handler
        log_handler = QLogHandler(self)
        log_handler.setFormatter(
            logging.Formatter("[%(levelname)s] %(name)s - %(message)s")
        )
        root_logger.addHandler(log_handler)
        log_handler.signal.log_str.connect(self.log_box.append)

        grid.addWidget(QLabel("Log:"), 5, 0, Qt.AlignTop)
        grid.addWidget(self.log_box, 5, 1)

        # setup the overall layout for the widget
        self.setLayout(grid)

        # debug shortcuts
        self.debug = QShortcut(QKeySequence("Shift+F8"), self)
        self.debug.activated.connect(self.toggle_debug)

    def access_reference(self):
        """Create thread to run the api

        When the search button is pressed, the contents is reset
        and a new thread is created to run the api object
        """

        self.reset_content()
        reference = self.ref_line.text()

        if reference:
            gui_logger.info(f"Search reference: {reference}")
            ref_thread = RefThread(reference, self.format_config, parent=self)
            ref_thread.response_obj.connect(self.output)
            ref_thread.start()

        else:
            gui_logger.warning(f"Please enter doi or arXiv ID")

    @Slot(object)
    def output(self, api_object):
        """Link the api object to GUI class

        The slot for the signal from access_reference
        store the api object to the GUI
        """
        self.api_object = api_object
        self.change_format()

    def change_format(self):
        """Test the group of button if it is checked"""
        ref_format = self.format_btns.checkedButton().text()
        self.output_box.clear()
        if self.api_object.status:
            gui_logger.debug(f"{ref_format} format")
            format_thread = FormatThread(
                self.api_object, ref_format, parent=self
            )
            format_thread.response_str.connect(self.update_output)
            format_thread.start()

    @Slot(str)
    def update_output(self, output_str):
        """Update output from stored parsed api object

        This is done by first checking which format button is clicked
        """
        self.output_box.setText(output_str)

    def reset_content(self):
        """Clear and reset the contents"""
        self.output_box.clear()
        self.log_box.clear()

    def copy(self):
        """Copy the output text"""
        clipboard = QApplication.clipboard()
        clipboard.setText(self.output_box.toPlainText())
        gui_logger.info("text copied to clipboard")

    def toggle_debug(self):
        """Secret short cut shift + F8 to switch to debug mode"""

        if root_logger.isEnabledFor(logging.DEBUG):
            gui_logger.info("switch to regular mode")
            root_logger.setLevel(logging.INFO)
        else:
            gui_logger.info("switch to debug mode")
            root_logger.setLevel(logging.DEBUG)

    def closeEvent(self, event):
        """Exit and wait for the thread to finish"""
        for thread in self.__threads:
            thread[0].quit()
            thread[0].wait()


class RefThread(QThread):
    """Reference parsing thread for the GUI."""

    response_obj = Signal(object)

    def __init__(self, reference, format_config, parent=None):
        super().__init__(parent)
        self.reference = reference
        self.format_config = format_config

    def run(self):
        """Emit the parsed api object"""
        self.response_obj.emit(RefAPI(self.reference, self.format_config))


class FormatThread(QThread):
    """Convert result to templated formats"""

    response_str = Signal(str)

    def __init__(self, api_object, ref_format, parent=None):
        super().__init__(parent)
        self.api_object = api_object
        self.ref_format = ref_format

    def run(self):
        """Emit the parsed ref_format string"""
        self.response_str.emit(self.api_object.render(self.ref_format))


class QLogSignal(QThread):
    """Qt signal for logging

    To create a thread safe logging, need to use signal-slot to pass the
    logging and formatting message
    """

    log_str = Signal(str)

    def __init__(self, parent=None):
        super().__init__(parent)


class QLogHandler(logging.Handler):
    """Custom handler to stream log to QTextEdit

    To have a thread safe to exit, need to define parent of the signal
    thread, here passed as the parent parameters.
    """

    def __init__(self, parent=None):
        super().__init__()
        self.signal = QLogSignal(parent)

    def emit(self, record):
        """Emit colored log record"""
        coloredlog = self.colorlog(record)
        self.signal.log_str.emit(coloredlog)

    def colorlog(self, record):
        """Set different color to error and warning levels

        The levels colored: warning - yellow, error - red
        and critical - purple

        """
        msg = self.format(record)
        color_format = "<span style='color:{}'>{}</span>"
        level_color = defaultdict(lambda: "black")
        level_color.update(
            {
                "WARNING": "DarkOrange",
                "ERROR": "red",
                "CRITICAL": "purple",
                "DEBUG": "SteelBlue",
            }
        )
        return color_format.format(level_color[record.levelname], msg)


def refparse_gui(formats):
    """main function call for GUI

    The function called by the commendline interface
    :param formats dict: dictionary of format configurations
    """
    parser = QApplication(sys.argv)
    font = QFont()
    font.setFamily("Arial")
    parser.setFont(font)
    parser.setStyle("fusion")
    main_parser = ParserGUI(formats)
    main_parser.show()
    sys.exit(parser.exec_())


refparse/utils.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""Utility functions for parser and Cheetah3 templat"""


from pylatexenc.latexencode import unicode_to_latex
from calendar import month_abbr, month_name
from titlecase import titlecase
import re
import bs4
from collections import defaultdict


class Empty(object):
    """Use to create empty object

    Used as an placeholder for bs4 tag object.
    string and get_text method return empty strin"""

    string = ""

    def get_text(self, *args, **kwargs):
        return ""


def get_attr(obj, path):
    """Get nested attribute

    :param obj object/namedtuple: object to get the attribute
    :param path str: nested attribute path, separated by '/'
    """

    try:
        for p in path.split("/"):
            obj = getattr(obj, p)
        return obj
    except AttributeError:
        return Empty()


def get_string(obj, path):
    """Get the string attribute of nested object

    With strip=True, text with html tags around is be stripped away
    its white space around
    :param obj object/namedtuple: object to get the attribute
    :param path str: nested attribute path, separated by '/'
    """
    try:
        return get_attr(obj, path).get_text(strip=True)
    except AttributeError:
        return ""


def html_convert(tag_element):
    """extract the html content of the tag element

    Extract the html of the tag element, and present it as close
    as the browser. Currently this function is limited.

    :param tag_element bs4.element.tag: tag to extract content
        This should be a soup object
    """

    _html_to_latex_tags = defaultdict(lambda: ("", ""))
    _html_to_latex_tags.update(
        {
            "i": ("\\textit{", "}"),
            "b": ("\\textbf{", "}"),
            "strong": ("\\textbf{", "}"),
            "em": ("\\emph{", "}"),
            "u": ("\\underline{", "}"),
            "sub": ("\\textsubscript{", "}"),
            "sup": ("\\textsuperscript{", "}"),
        }
    )

    if tag_element is None:
        return "", "", ""

    content = []
    content_html = []
    content_latex = []
    tag_pattern = re.compile(r"<\w+\s*\w*>(.+)</(\w+)>")

    for ele in tag_element:
        if not isinstance(ele, bs4.element.Tag):
            str_ = re.sub(r"([\n].+[\n])\s+", r"\1", ele)
            str_ = re.sub(r"\s+", " ", str_.replace("\n", " "))
            content.append(str_)
            content_html.append(str_)
            content_latex.append(str_)
        else:
            subtag = re.match(tag_pattern, str(ele))
            text = subtag.group(1)
            tag = subtag.group(2)

            content.append(text)
            content_html.append(str(ele))
            latex = "{latex[0]}{text}{latex[1]}".format(
                latex=_html_to_latex_tags[tag], text=text,
            )
            content_latex.append(latex)
    return (
        "".join(content).strip(),
        "".join(content_latex).strip(),
        "".join(content_html).strip(),
    )


class Filters:
    """Filter used for Cheetah Template engine"""

    @classmethod
    def list(cls, *args, **kwargs):
        return list(*args, **kwargs)

    @classmethod
    def map(cls, *args, **kwargs):
        return map(*args, **kwargs)

    @classmethod
    def titlecase(cls, text):
        """A wrapper for titlecase function"""
        return titlecase(text)

    @classmethod
    def month_abbr(cls, month):
        """A wrapper for calendar.month_abbr

        :param month str: month string or int, empty string will
            be converted to 0, and return an empty string
        """
        month = int(month) if month else 0
        return month_abbr[month]

    @classmethod
    def month_name(cls, month):
        """A wrapper for calendar.month_name

        :param month str: month string or int, empty string will
            be converted to 0, and return an empty string
        """
        month = int(month) if month else 0
        return month_name[month]

    @classmethod
    def unicode_to_latex(cls, text):
        """A wrapper for calendar.month_name"""
        return unicode_to_latex(text)


refparse/parser.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""Reference parser for different APIs"""


from refparse.utils import get_attr, get_string, html_convert
from bs4 import BeautifulSoup

import requests
import logging
from collections import defaultdict
import abc
from datetime import datetime
import re


class ParserBase(abc.ABC):
    """Abstract method for parsers

    If the attribute is not found, a empty string will be returned
    """

    REFNAME: str
    REF_URL: str
    QUERY_URL: str
    HEADER: dict

    def __init__(self, reference):

        self.log = logging.getLogger(self.__class__.__name__)
        self.query_url = self.QUERY_URL.format(reference)
        self.ok, self.text = self.request_text(self.query_url)
        self.parsed = defaultdict(str)

        if self.ok:
            # needs to use xml, abstract does not show up with lxml
            self.soup = BeautifulSoup(self.text, "xml")
            self.parsed[self.REFNAME] = reference
            self.parsed["reference"] = reference
            self.parsed["ref_type"] = self.REFNAME.replace(" ", "_")
            self.parsed["url"] = self.REF_URL.format(reference)
            # parse api
            self.parsed.update(self.parse_api(self.soup))

    def request_text(self, url):

        r = requests.get(
            url, headers={"Accept": "application/vnd.crossref.unixsd+xml"}
        )
        if r.ok:
            self.log.info(f"{self.REFNAME} found")
        elif r.status_code == 404 or r.status_code == 400:
            self.log.error(f"Incorrect {self.REFNAME}")
        elif r.status_code == 504:
            self.log.error(f"Gateway timeout, please try again")
        r.encoding = "utf-8"
        return r.ok, r.text

    @abc.abstractmethod
    def parse_api(self, soup):
        """The main function to parse api

        The method is required. This should be replaced for parsers
        """
        return {}


class CrossRefParser(ParserBase):

    REFNAME = "doi"
    REF_URL = "http://doi.org/{}"
    QUERY_URL = "http://dx.doi.org/{}"
    HEADER = {"Accept": "application/vnd.crossref.unixsd+xml"}

    def parse_api(self, soup):
        pdict = {}

        pdict["has_publication"] = True
        journal_meta = soup.journal_metadata
        pdict["journal_full_title"] = get_string(journal_meta, "full_title")
        pdict["journal_abbrev_title"] = get_string(
            journal_meta, "abbrev_title"
        )

        article_meta = soup.journal_article

        author = []
        author_tag = get_attr(article_meta, "contributors")
        for name in author_tag.find_all("person_name"):
            author.append([name.surname.string, name.given_name.string])
        pdict["author"] = author

        (
            pdict["title"],
            pdict["title_latex"],
            pdict["title_html"],
        ) = html_convert(get_attr(article_meta, "titles/title"))

        pdict["abstract"] = get_string(article_meta, "abstract")

        pub_online = article_meta.find(
            "publication_date", {"media_type": "online"}
        )
        pdict["online_year"] = get_string(pub_online, "year")
        pdict["online_month"] = get_string(pub_online, "month")
        pdict["online_day"] = get_string(pub_online, "day")

        pub_print = article_meta.find(
            "publication_date", {"media_type": "print"}
        )
        if pub_print:
            self.log.info("print version found")
            pdict["has_print"] = True
            pdict["print_year"] = get_string(pub_online, "year")
            pdict["print_month"] = get_string(pub_online, "month")
            pdict["print_day"] = get_string(pub_online, "day")

            first_page = get_string(soup, "pages/first_page")
            last_page = get_string(soup, "pages/last_page")
            pdict["pages"] = (
                [first_page, last_page] if last_page else [first_page]
            )

            issue_meta = soup.journal_issue
            pdict["volume"] = get_string(issue_meta, "journal_volume/volume")
            pdict["issue"] = get_string(issue_meta, "issue")
        else:
            pdict["has_print"] = False
        return pdict


class arXivParser(ParserBase):
    REF_URL = "http://arxiv.org/{}"
    QUERY_URL = "http://export.arxiv.org/api/query?id_list={}"
    REFNAME = "arXiv ID"
    HEADER = {}

    def search_doi(self, soup):
        """Check if the article has doi"""
        doi_tag = soup.find("link", {"title": "doi"})
        if doi_tag:
            self.log.warning(f"article has doi: {doi_tag['href']}")

    def parse_api(self, soup):
        """Parse the article information"""
        pdict = {}
        pdict["has_publication"] = False
        pdict["has_print"] = False
        self.search_doi(soup)

        article_meta = soup.entry
        # remove unnecessary line break
        pdict["abstract"] = get_string(article_meta, "summary").replace(
            "\n", " "
        )
        print(repr(article_meta.summary.get_text(strip=True)))
        # sometimes the arXiv article title has unnecessary linebreak
        pdict["title"] = get_string(article_meta, "title").replace("\n ", "")
        pdict["title_latex"] = pdict["title"]

        pub_date = datetime.strptime(
            article_meta.updated.string, "%Y-%m-%dT%H:%M:%SZ"
        )
        pdict["online_year"] = str(pub_date.year)
        pdict["online_month"] = str(pub_date.month)
        pdict["online_day"] = str(pub_date.day)

        author = []
        for name in article_meta.find_all("name"):
            name_ = re.match(r"([\s\S]+) (\w+)", name.string)
            author.append([name_.group(2), name_.group(1)])
        pdict["author"] = author
        return pdict
